### DISCUSSION TITLE
How to change Logger

### DISCUSSION THREAD

**Discussion by Northward712** [ergo-services/ergo#151] [Q&A] [answered]
I want use self Logger 
How to change Logger in ergo?

**Comment by halturin** [comment]
Thanks for the question. Golang has no Logger interface yet (they will probably add it soon https://github.com/golang/go/issues/56345). Until then, there is no standard way to do that. What is the reason for having it in your program?

**Reply by Northward712** [reply]
I think you can add a custom interface Logger to the lib.tools file.
e.g. 
`type Logger interface {
Warningf(format string,args...any)
    Infof(format string,args...any)
 Debugf(format string,args...any)
}`
And then add a function.

e.g.
func SetLogger(log Logger) { defaultLog = log }

In this way, there is no need to use the Logger of the standard library, so that users can customize the Logger and avoid the mismatch between the style of their own program logs when using ergo.

**Reply by Northward712** [reply]
I use Zap in my program

In many cases, the log printed by your framework is not consistent with the log structure I encapsulated.

In addition, in a production environment, you may need to print the log in json format and write to the file, but in this case, the Logger of the standard library will have performance problems and a separate set of code for the standard library Logger will be required.

**Reply by Northward712** [reply]
e.g.
![截屏2023-03-29 09 01 23](https://user-images.githubusercontent.com/37608723/228399747-1d1e5e82-f254-4d6d-bfbc-975c79278dd2.png)
Above is my journal, and below is yours.

This kind of log printing makes it impossible for me to use the standard format for alarms during log monitoring and error alarms.

**Reply by Northward712** [reply]
Thank you for your reply.

**Comment by halturin** [comment]
moving it to the discussion

**Comment by halturin** [comment, marked as answer]
the next version 3.0 will get a flexible logging subsystem. you will be able to add a custom logger or use the process as a logger. you will also be able to set the logging level for the node and the process individually. here is little example how it looks like
```golang
func (a *MyActor) Init(args ...any) error {
    ...
    a.Log().Info("starting with name - %s", a.Name())
    a.Log().SetLevel(gen.LogLevelDebug)
    return nil
}
...
pid, err := node.SpawnRegister("example", fabricMyActor, gen.ProcessOptions{})
node.Log().Info("started MyActor with PID %s", pid)
node.Log().SetLevel(gen.LogLevelDebug)
...
1686217717548243545 [info] <90A29F11.0.1001>: starting with name - 'example'
1686217717548282816 [info] 90A29F11: started MyActor with PID <90A29F11.0.1001>
```
PS: its done now, just wait for the release

**Reply by Northward712** [reply]
Thank you for not knowing if there is a specific release date.

**Reply by halturin** [reply]
The release is planned for this fall.
